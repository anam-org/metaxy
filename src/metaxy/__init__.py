from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
