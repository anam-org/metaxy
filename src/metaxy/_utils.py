import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
