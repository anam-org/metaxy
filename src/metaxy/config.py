"""Configuration system for Metaxy using pydantic-settings."""

from pathlib import Path
from typing import TYPE_CHECKING, Any, TypeVar, overload

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Fallback for Python 3.10

from pydantic import Field as PydanticField
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
)

if TYPE_CHECKING:
    from metaxy.metadata_store.base import MetadataStore
    from metaxy.models.feature import FeatureRegistry

T = TypeVar("T")


def load_registry(registry_path: str) -> "FeatureRegistry":
    """Load registry from import path.

    Args:
        registry_path: Import path like "myapp.features:my_registry"

    Returns:
        FeatureRegistry instance

    Raises:
        ValueError: If path is invalid
        ImportError: If module or registry not found
        TypeError: If loaded object is not a FeatureRegistry
    """
    from metaxy.models.feature import FeatureRegistry

    if ":" not in registry_path:
        raise ValueError(
            f"Invalid registry path: {registry_path}. "
            f"Expected format: 'module.path:registry_name'"
        )

    module_path, registry_name = registry_path.split(":", 1)

    try:
        module = __import__(module_path, fromlist=[registry_name])
        registry = getattr(module, registry_name)
    except (ImportError, AttributeError) as e:
        raise ImportError(f"Failed to load registry from {registry_path}: {e}") from e

    if not isinstance(registry, FeatureRegistry):
        raise TypeError(
            f"Registry at {registry_path} is not a FeatureRegistry instance. "
            f"Got: {type(registry)}"
        )

    return registry


class TomlConfigSettingsSource(PydanticBaseSettingsSource):
    """Custom settings source for TOML configuration files.

    Auto-discovers configuration in this order:
    1. Explicit file path if provided
    2. metaxy.toml in current directory (preferred)
    3. pyproject.toml [tool.metaxy] section (fallback)
    4. No config (returns empty dict)
    """

    def __init__(self, settings_cls: type[BaseSettings], toml_file: Path | None = None):
        super().__init__(settings_cls)
        self.toml_file = toml_file or self._discover_config_file()
        self.toml_data = self._load_toml()

    def _discover_config_file(self) -> Path | None:
        """Auto-discover config file."""
        # Prefer metaxy.toml
        if Path("metaxy.toml").exists():
            return Path("metaxy.toml")

        # Fallback to pyproject.toml
        if Path("pyproject.toml").exists():
            return Path("pyproject.toml")

        return None

    def _load_toml(self) -> dict[str, Any]:
        """Load TOML file and extract metaxy config."""
        if self.toml_file is None:
            return {}

        with open(self.toml_file, "rb") as f:
            data = tomllib.load(f)

        # Extract [tool.metaxy] from pyproject.toml or root from metaxy.toml
        if self.toml_file.name == "pyproject.toml":
            return data.get("tool", {}).get("metaxy", {})
        else:
            return data

    def get_field_value(self, field: Any, field_name: str) -> tuple[Any, str, bool]:
        """Get field value from TOML data."""
        field_value = self.toml_data.get(field_name)
        return field_value, field_name, False

    def __call__(self) -> dict[str, Any]:
        """Return all settings from TOML."""
        return self.toml_data


class StoreConfig(BaseSettings):
    """Configuration for a single metadata store.

    Structure:
        type: Full import path to store class
        config: Dict of all configuration (including fallback_stores)

    Example:
        >>> config = StoreConfig(
        ...     type="metaxy_delta.DeltaMetadataStore",
        ...     config={
        ...         "table_uri": "s3://bucket/metadata",
        ...         "region": "us-west-2",
        ...         "fallback_stores": ["prod"],
        ...     }
        ... )
    """

    model_config = SettingsConfigDict(
        extra="forbid",  # Only type and config fields allowed
    )

    # Store class (full import path)
    type: str

    # Store configuration (all kwargs for __init__)
    # This includes fallback_stores, table_uri, db_path, storage_options, etc.
    config: dict[str, Any] = PydanticField(default_factory=dict)


class MetaxyConfig(BaseSettings):
    """Main Metaxy configuration.

    Loads from:
    1. TOML file (metaxy.toml or pyproject.toml [tool.metaxy])
    2. Environment variables (METAXY_*)
    3. Init arguments

    Priority: init > env vars > TOML

    Example:
        >>> # Auto-discover config
        >>> config = MetaxyConfig.load()
        >>>
        >>> # Get store instance
        >>> store = config.get_store("prod")
        >>>
        >>> # Override via env var
        >>> # METAXY_STORE=staging METAXY_REGISTRY=myapp.features:my_registry
        >>> config = MetaxyConfig.load()
        >>> store = config.get_store()  # Uses staging with custom registry
    """

    model_config = SettingsConfigDict(
        env_prefix="METAXY_",
        env_nested_delimiter="__",
    )

    # Store to use
    store: str = "dev"

    # Named store configurations
    stores: dict[str, StoreConfig] = PydanticField(default_factory=dict)

    # Migrations directory
    migrations_dir: str = "metaxy/migrations"

    # Entrypoints to load (list of module paths)
    entrypoints: list[str] = PydanticField(default_factory=list)

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        """Customize settings sources: init → env → TOML.

        Priority (first wins):
        1. Init arguments
        2. Environment variables
        3. TOML file
        """
        toml_settings = TomlConfigSettingsSource(settings_cls)
        return (init_settings, env_settings, toml_settings)

    @classmethod
    def load(
        cls, config_file: str | Path | None = None, *, search_parents: bool = True
    ) -> "MetaxyConfig":
        """Load config with auto-discovery and parent directory search.

        Args:
            config_file: Optional config file path (overrides auto-discovery)
            search_parents: Search parent directories for config file (default: True)

        Returns:
            Loaded config (TOML + env vars merged)

        Example:
            >>> # Auto-discover with parent search
            >>> config = MetaxyConfig.load()
            >>>
            >>> # Explicit file
            >>> config = MetaxyConfig.load("custom.toml")

            >>> # Auto-discover without parent search
            >>> config = MetaxyConfig.load(search_parents=False)
        """
        # Search for config file if not explicitly provided
        if config_file is None and search_parents:
            config_file = cls._discover_config_with_parents()

        # For explicit file, temporarily patch the TomlConfigSettingsSource
        # to use that file, then use normal instantiation
        # This ensures env vars still work

        if config_file:
            # Create a custom settings source class for this file
            toml_path = Path(config_file)

            class CustomTomlSource(TomlConfigSettingsSource):
                def __init__(self, settings_cls: type[BaseSettings]):
                    # Skip auto-discovery, use explicit file
                    super(TomlConfigSettingsSource, self).__init__(settings_cls)
                    self.toml_file = toml_path
                    self.toml_data = self._load_toml()

            # Customize sources to use custom TOML file
            original_method = cls.settings_customise_sources

            @classmethod  # type: ignore[misc]
            def custom_sources(
                cls_inner,
                settings_cls,
                init_settings,
                env_settings,
                dotenv_settings,
                file_secret_settings,
            ):
                toml_settings = CustomTomlSource(settings_cls)
                return (init_settings, env_settings, toml_settings)

            # Temporarily replace method
            cls.settings_customise_sources = custom_sources  # type: ignore[assignment]
            config = cls()
            cls.settings_customise_sources = original_method  # type: ignore[method-assign]
        else:
            # Use default sources (auto-discovery + env vars)
            config = cls()

        # Load entrypoints if configured (do this after config is created)
        if config.entrypoints:
            from metaxy.entrypoints import load_config_entrypoints

            load_config_entrypoints(config.entrypoints)

        return config

    @staticmethod
    def _discover_config_with_parents() -> Path | None:
        """Discover config file by searching current and parent directories.

        Searches for metaxy.toml or pyproject.toml in current directory,
        then iteratively searches parent directories.

        Returns:
            Path to config file if found, None otherwise
        """
        current = Path.cwd()

        while True:
            # Check for metaxy.toml (preferred)
            metaxy_toml = current / "metaxy.toml"
            if metaxy_toml.exists():
                return metaxy_toml

            # Check for pyproject.toml
            pyproject_toml = current / "pyproject.toml"
            if pyproject_toml.exists():
                return pyproject_toml

            # Move to parent
            parent = current.parent
            if parent == current:
                # Reached root
                break
            current = parent

        return None

    def get_store(
        self,
        name: str | None = None,
    ) -> "MetadataStore":
        """Instantiate metadata store by name.

        Args:
            name: Store name (uses config.store if None)

        Returns:
            Instantiated metadata store

        Raises:
            ValueError: If store name not found in config, or if fallback stores
                have different hash algorithms than the parent store
            ImportError: If store class cannot be imported

        Example:
            >>> config = MetaxyConfig.load()
            >>> store = config.get_store("prod")
            >>>
            >>> # Use default store
            >>> store = config.get_store()
        """
        from metaxy.data_versioning.hash_algorithms import HashAlgorithm

        if len(self.stores) == 0:
            raise ValueError(
                "No Metaxy stores available. They should be configured in metaxy.toml|pyproject.toml or via environment variables."
            )

        name = name or self.store

        if name not in self.stores:
            raise ValueError(
                f"Store '{name}' not found in config. "
                f"Available stores: {list(self.stores.keys())}"
            )

        store_config = self.stores[name]

        # Import store class
        store_class = self._import_class(store_config.type)

        # Extract configuration
        config_copy = store_config.config.copy()
        fallback_store_names = config_copy.pop("fallback_stores", [])

        # Get hash_algorithm from config (if specified) and convert to enum
        configured_hash_algorithm = config_copy.get("hash_algorithm")
        if configured_hash_algorithm is not None:
            # Convert string to enum if needed
            if isinstance(configured_hash_algorithm, str):
                configured_hash_algorithm = HashAlgorithm(configured_hash_algorithm)
                config_copy["hash_algorithm"] = configured_hash_algorithm
        else:
            # Use default
            configured_hash_algorithm = HashAlgorithm.XXHASH64
            config_copy["hash_algorithm"] = configured_hash_algorithm

        # Build fallback stores recursively
        fallback_stores = []
        for fallback_name in fallback_store_names:
            fallback_store = self.get_store(fallback_name)
            fallback_stores.append(fallback_store)

        # Instantiate store with config + fallback_stores
        store = store_class(
            fallback_stores=fallback_stores,
            **config_copy,
        )

        # Verify the store actually uses the hash algorithm we configured
        # (in case a store subclass overrides the default or ignores the parameter)
        if store.hash_algorithm != configured_hash_algorithm:
            raise ValueError(
                f"Store '{name}' ({store_class.__name__}) was configured with "
                f"hash_algorithm='{configured_hash_algorithm.value}' but is using "
                f"'{store.hash_algorithm.value}'. The store class may have overridden "
                f"the hash algorithm. All stores must use the same hash algorithm."
            )

        return store

    @staticmethod
    def _import_class(class_path: str) -> type:
        """Import class from module path.

        Args:
            class_path: Full import path like "metaxy.metadata_store.InMemoryMetadataStore"

        Returns:
            Imported class

        Raises:
            ImportError: If module or class not found
        """
        module_path, class_name = class_path.rsplit(".", 1)
        module = __import__(module_path, fromlist=[class_name])
        return getattr(module, class_name)

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: type[T]) -> T: ...

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: None = None) -> Any: ...

    @staticmethod
    def _import_object(
        import_path: str, expected_type: type[T] | None = None
    ) -> T | Any:
        """Import object from module:object path.

        Args:
            import_path: Import path like "module.path:object_name"
            expected_type: Optional type to validate against

        Returns:
            Imported object (typed as expected_type if provided)

        Raises:
            ValueError: If path format is invalid
            ImportError: If module or object not found
            TypeError: If object doesn't match expected_type
        """
        if ":" not in import_path:
            raise ValueError(
                f"Invalid import path: {import_path}. "
                f"Expected format: 'module.path:object_name'"
            )

        module_path, object_name = import_path.split(":", 1)

        try:
            module = __import__(module_path, fromlist=[object_name])
            obj = getattr(module, object_name)
        except (ImportError, AttributeError) as e:
            raise ImportError(f"Failed to load object from {import_path}: {e}") from e

        if expected_type is not None and not isinstance(obj, expected_type):
            type_name = getattr(expected_type, "__name__", str(expected_type))
            raise TypeError(
                f"Object at {import_path} is not a {type_name} instance. "
                f"Got: {type(obj)}"
            )

        return obj
