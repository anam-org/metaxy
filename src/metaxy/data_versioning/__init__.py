"""Data versioning module for sample-level data version calculation."""

from metaxy.data_versioning.calculators import (
    DataVersionCalculator,
    PolarsDataVersionCalculator,
)
from metaxy.data_versioning.diff import (
    DiffResult,
    MetadataDiffResolver,
    PolarsDiffResolver,
)
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.data_versioning.joiners import PolarsJoiner, UpstreamJoiner

__all__ = [
    "HashAlgorithm",
    "UpstreamJoiner",
    "PolarsJoiner",
    "DataVersionCalculator",
    "PolarsDataVersionCalculator",
    "DiffResult",
    "MetadataDiffResolver",
    "PolarsDiffResolver",
]
