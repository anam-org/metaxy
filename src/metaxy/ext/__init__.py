"""Integrations with third-party software."""
