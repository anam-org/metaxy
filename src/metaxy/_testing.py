import importlib
import sys
import tempfile
from functools import cached_property
from pathlib import Path
from typing import Any

from metaxy import (
    FeatureSpec,
)
from metaxy.config import MetaxyConfig
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.metadata_store.base import MetadataStore
from metaxy.models.feature import FeatureGraph


class TempFeatureModule:
    """Helper to create temporary Python modules with feature definitions.

    This allows features to be importable by historical graph reconstruction.
    The same import path (e.g., 'temp_features.Upstream') can be used across
    different feature versions by overwriting the module file.
    """

    def __init__(self, module_name: str = "temp_test_features"):
        self.temp_dir = tempfile.mkdtemp(prefix="metaxy_test_")
        self.module_name = module_name
        self.module_path = Path(self.temp_dir) / f"{module_name}.py"

        # Add to sys.path so module can be imported
        sys.path.insert(0, self.temp_dir)

    def write_features(self, feature_specs: dict[str, FeatureSpec]):
        """Write feature classes to the module file.

        Args:
            feature_specs: Dict mapping class names to FeatureSpec objects
        """
        code_lines = [
            "# Auto-generated test feature module",
            "from metaxy import Feature, FeatureSpec, FieldSpec, FieldKey, FeatureDep, FeatureKey, FieldDep, SpecialFieldDep",
            "from metaxy.models.feature import FeatureGraph",
            "",
            "# Use a dedicated graph for this temp module",
            "_graph = FeatureGraph()",
            "",
        ]

        for class_name, spec in feature_specs.items():
            # Generate the spec definition
            spec_dict = spec.model_dump(mode="python")
            spec_repr = self._generate_spec_repr(spec_dict)

            code_lines.extend(
                [
                    f"# Define {class_name} in the temp graph context",
                    "with _graph.use():",
                    f"    class {class_name}(",
                    "        Feature,",
                    f"        spec={spec_repr}",
                    "    ):",
                    "        pass",
                    "",
                ]
            )

        # Write the file
        self.module_path.write_text("\n".join(code_lines))

        # Reload module if it was already imported
        if self.module_name in sys.modules:
            importlib.reload(sys.modules[self.module_name])

    def _generate_spec_repr(self, spec_dict: dict[str, Any]) -> str:
        """Generate FeatureSpec constructor call from dict."""
        # This is a simple representation - could be made more robust
        parts = []

        # key
        key = spec_dict["key"]
        parts.append(f"key=FeatureKey({key!r})")

        # deps
        deps = spec_dict.get("deps")
        if deps is None:
            parts.append("deps=None")
        else:
            deps_repr = [f"FeatureDep(key=FeatureKey({d['key']!r}))" for d in deps]
            parts.append(f"deps=[{', '.join(deps_repr)}]")

        # fields
        fields = spec_dict.get("fields", [])
        if fields:
            field_reprs = []
            for c in fields:
                c_parts = [
                    f"key=FieldKey({c['key']!r})",
                    f"code_version={c['code_version']}",
                ]

                # Handle deps
                deps_val = c.get("deps")
                if deps_val == "__METAXY_ALL_DEP__":
                    c_parts.append("deps=SpecialFieldDep.ALL")
                elif isinstance(deps_val, list) and deps_val:
                    # Field deps (list of FieldDep)
                    cdeps: list[str] = []  # type: ignore[misc]
                    for cd in deps_val:
                        fields_val = cd.get("fields")
                        if fields_val == "__METAXY_ALL_DEP__":
                            cdeps.append(  # type: ignore[arg-type]
                                f"FieldDep(feature_key=FeatureKey({cd['feature_key']!r}), fields=SpecialFieldDep.ALL)"
                            )
                        else:
                            # Build list of FieldKey objects
                            field_keys = [f"FieldKey({k!r})" for k in fields_val]
                            cdeps.append(
                                f"FieldDep(feature_key=FeatureKey({cd['feature_key']!r}), fields=[{', '.join(field_keys)}])"
                            )
                    c_parts.append(f"deps=[{', '.join(cdeps)}]")

                field_reprs.append(f"FieldSpec({', '.join(c_parts)})")  # type: ignore[arg-type]

            parts.append(f"fields=[{', '.join(field_reprs)}]")

        return f"FeatureSpec({', '.join(parts)})"

    @property
    def graph(self) -> FeatureGraph:
        """Get the FeatureGraph from the temp module.

        Returns:
            The _graph instance from the imported module
        """
        # Import the module to get its _graph
        module = importlib.import_module(self.module_name)
        return module._graph

    def cleanup(self):
        """Remove temp directory and module from sys.path.

        NOTE: Don't call this until the test session is completely done,
        as historical graph loading may need to import from these modules.
        """
        if self.temp_dir in sys.path:
            sys.path.remove(self.temp_dir)

        # Remove from sys.modules
        if self.module_name in sys.modules:
            del sys.modules[self.module_name]

        # Delete temp directory
        import shutil

        shutil.rmtree(self.temp_dir, ignore_errors=True)


def assert_all_results_equal(results: dict[str, Any], snapshot=None) -> None:
    """Compare all results from different store type combinations.

    Ensures all variants produce identical results, then optionally snapshots all results.

    Args:
        results: Dict mapping store_type to result data
        snapshot: Optional syrupy snapshot fixture to record all results

    Raises:
        AssertionError: If any variants produce different results
    """
    if not results:
        return

    # Get all result values as a list
    all_results = list(results.items())
    reference_key, reference_result = all_results[0]

    # Compare each result to the reference
    for key, result in all_results[1:]:
        assert result == reference_result, (
            f"{key} produced different results than {reference_key}:\n"
            f"Expected: {reference_result}\n"
            f"Got: {result}"
        )

    # Snapshot ALL results if snapshot provided
    if snapshot is not None:
        assert results == snapshot


class HashAlgorithmCases:
    """Test cases for different hash algorithms."""

    def case_xxhash64(self) -> HashAlgorithm:
        """xxHash64 algorithm."""
        return HashAlgorithm.XXHASH64

    def case_xxhash32(self) -> HashAlgorithm:
        """xxHash32 algorithm."""
        return HashAlgorithm.XXHASH32

    def case_wyhash(self) -> HashAlgorithm:
        """WyHash algorithm."""
        return HashAlgorithm.WYHASH

    def case_sha256(self) -> HashAlgorithm:
        """SHA256 algorithm."""
        return HashAlgorithm.SHA256

    def case_md5(self) -> HashAlgorithm:
        """MD5 algorithm."""
        return HashAlgorithm.MD5


class TempMetaxyProject:
    """Helper for creating temporary Metaxy projects.

    Provides a context manager API for dynamically creating feature modules
    and running CLI commands with proper entrypoint configuration.

    Example:
        >>> project = TempMetaxyProject(tmp_path)
        >>>
        >>> def features():
        ...     from metaxy import Feature, FeatureSpec, FeatureKey, FieldSpec, FieldKey
        ...
        ...     class MyFeature(Feature, spec=FeatureSpec(
        ...         key=FeatureKey(["my_feature"]),
        ...         deps=None,
        ...         fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)]
        ...     )):
        ...         pass
        >>>
        >>> with project.with_features(features):
        ...     result = project.run_cli("graph", "push")
        ...     assert result.returncode == 0
    """

    def __init__(self, tmp_path: Path):
        """Initialize a temporary Metaxy project.

        Args:
            tmp_path: Temporary directory path (usually from pytest tmp_path fixture)
        """
        self.project_dir = tmp_path
        self.project_dir.mkdir(exist_ok=True)
        self._feature_modules: list[str] = []
        self._write_config()

    def _write_config(self):
        """Write basic metaxy.toml with DuckDB store configuration."""
        dev_db_path = self.project_dir / "metadata.duckdb"
        staging_db_path = self.project_dir / "metadata_staging.duckdb"
        config_content = f'''store = "dev"

[stores.dev]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.dev.config]
database = "{dev_db_path}"

[stores.staging]
type = "metaxy.metadata_store.duckdb.DuckDBMetadataStore"

[stores.staging.config]
database = "{staging_db_path}"
'''
        (self.project_dir / "metaxy.toml").write_text(config_content)

    def with_features(self, features_func, module_name: str | None = None):
        """Context manager that sets up features for the duration of the block.

        Extracts source code from features_func (skipping the function definition line),
        writes it to a Python module file, and tracks it for METAXY_ENTRYPOINTS__N
        environment variable configuration.

        Args:
            features_func: Function containing feature class definitions.
                All imports must be inside the function body.
            module_name: Optional module name. If not provided, generates
                "features_N" based on number of existing modules.

        Yields:
            str: The module name that was created

        Example:
            >>> def my_features():
            ...     from metaxy import Feature, FeatureSpec, FeatureKey
            ...
            ...     class MyFeature(Feature, spec=...):
            ...         pass
            >>>
            >>> with project.with_features(my_features) as module:
            ...     print(module)  # "features_0"
            ...     result = project.run_cli("graph", "push")
        """
        import inspect
        import textwrap
        from contextlib import contextmanager

        @contextmanager
        def _context():
            # Generate module name if not provided
            nonlocal module_name
            if module_name is None:
                module_name = f"features_{len(self._feature_modules)}"

            # Extract source code from function
            source = inspect.getsource(features_func)

            # Remove function definition line and dedent
            lines = source.split("\n")
            # Find the first line that's not a decorator or function def
            body_start = 0
            for i, line in enumerate(lines):
                if line.strip().startswith("def ") and ":" in line:
                    body_start = i + 1
                    break

            body_lines = lines[body_start:]
            dedented = textwrap.dedent("\n".join(body_lines))

            # Write to file in project directory
            feature_file = self.project_dir / f"{module_name}.py"
            feature_file.write_text(dedented)

            # Track this module
            self._feature_modules.append(module_name)

            try:
                yield module_name
            finally:
                # Cleanup: remove from tracking (file stays for debugging)
                if module_name in self._feature_modules:
                    self._feature_modules.remove(module_name)

        return _context()

    def run_cli(
        self, *args, check: bool = True, env: dict[str, str] | None = None, **kwargs
    ):
        """Run CLI command with current feature modules loaded.

        Automatically sets METAXY_ENTRYPOINT_0, METAXY_ENTRYPOINT_1, etc.
        based on active with_features() context managers.

        Args:
            *args: CLI command arguments (e.g., "graph", "push")
            check: If True (default), raises CalledProcessError on non-zero exit
            env: Optional dict of additional environment variables
            **kwargs: Additional arguments to pass to subprocess.run()

        Returns:
            subprocess.CompletedProcess: Result of the CLI command

        Raises:
            subprocess.CalledProcessError: If check=True and command fails

        Example:
            >>> result = project.run_cli("graph", "history", "--limit", "5")
            >>> print(result.stdout)
        """
        import os
        import subprocess

        # Start with current environment
        cmd_env = os.environ.copy()

        # Add project directory to PYTHONPATH so modules can be imported
        pythonpath = str(self.project_dir)
        if "PYTHONPATH" in cmd_env:
            pythonpath = f"{pythonpath}{os.pathsep}{cmd_env['PYTHONPATH']}"
        cmd_env["PYTHONPATH"] = pythonpath

        # Set entrypoints for all tracked modules
        # Use METAXY_ENTRYPOINT_0, METAXY_ENTRYPOINT_1, etc. (single underscore for list indexing)
        for idx, module_name in enumerate(self._feature_modules):
            cmd_env[f"METAXY_ENTRYPOINT_{idx}"] = module_name

        # Apply additional env overrides
        if env:
            cmd_env.update(env)

        # Run CLI command
        result = subprocess.run(
            [sys.executable, "-m", "metaxy.cli.app", *args],
            cwd=str(self.project_dir),
            capture_output=True,
            text=True,
            env=cmd_env,
            check=check,
            **kwargs,
        )

        return result

    @property
    def entrypoints(self):
        return [f"METAXY_ENTRYPOINT_{idx}" for idx in range(len(self._feature_modules))]

    @property
    def graph(self) -> FeatureGraph:
        """Load features from the project's feature modules into a graph.

        Returns:
            FeatureGraph with all features from tracked modules loaded
        """
        import importlib
        import sys

        graph = FeatureGraph()

        # Ensure project dir is in sys.path
        project_dir_str = str(self.project_dir)
        was_in_path = project_dir_str in sys.path
        if not was_in_path:
            sys.path.insert(0, project_dir_str)

        try:
            with graph.use():
                # Import feature modules directly
                for module_name in self._feature_modules:
                    # Import or reload the module
                    if module_name in sys.modules:
                        importlib.reload(sys.modules[module_name])
                    else:
                        importlib.import_module(module_name)
        finally:
            # Clean up sys.path if we added it
            if not was_in_path and project_dir_str in sys.path:
                sys.path.remove(project_dir_str)

        return graph

    @cached_property
    def config(self) -> MetaxyConfig:
        return MetaxyConfig.load(self.project_dir / "metaxy.toml")

    @cached_property
    def stores(self) -> dict[str, MetadataStore]:
        return {k: self.config.get_store(k) for k in self.config.stores}
