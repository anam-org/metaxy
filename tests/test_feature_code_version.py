"""Tests for Feature.code_version."""

from __future__ import annotations

from uuid import uuid4

from hypothesis import given
from hypothesis import strategies as st

from metaxy import Feature, FeatureDep, FeatureKey, FeatureSpec, FieldKey, FieldSpec


def _build_feature(
    *,
    key_suffix: str,
    fields: list[FieldSpec],
    deps: list[FeatureDep] | None = None,
) -> type[Feature]:
    feature_key = FeatureKey(["code_version", key_suffix])

    class _DynamicFeature(
        Feature,
        spec=FeatureSpec(
            key=feature_key,
            deps=deps,
            fields=fields,
        ),
    ):
        pass

    return _DynamicFeature


def test_code_version_single_field(graph) -> None:
    feature_cls = _build_feature(
        key_suffix="single_field",
        fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
    )

    value = feature_cls.code_version

    assert len(value) == 64
    assert all(c in "0123456789abcdef" for c in value)


def test_code_version_multiple_fields_order_invariant(graph) -> None:
    field_items = [
        (FieldKey(["frames"]), 1),
        (FieldKey(["audio"]), 2),
        (FieldKey(["metadata"]), 3),
    ]

    def build_fields(items: list[tuple[FieldKey, int]]) -> list[FieldSpec]:
        return [
            FieldSpec(key=key, code_version=code_version) for key, code_version in items
        ]

    ordered_feature = _build_feature(
        key_suffix="multi_ordered",
        fields=build_fields(field_items),
    )
    reversed_feature = _build_feature(
        key_suffix="multi_reversed",
        fields=build_fields(list(reversed(field_items))),
    )

    assert ordered_feature.code_version == reversed_feature.code_version


def test_code_version_changes_when_field_changes(graph) -> None:
    feature_v1 = _build_feature(
        key_suffix="field_change_v1",
        fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
    )
    feature_v2 = _build_feature(
        key_suffix="field_change_v2",
        fields=[FieldSpec(key=FieldKey(["default"]), code_version=2)],
    )

    assert feature_v1.code_version != feature_v2.code_version


def test_code_version_ignores_dependency_versions(graph) -> None:
    parent_key = FeatureKey(["code_version", "parent"])

    class ParentV1(
        Feature,
        spec=FeatureSpec(
            key=parent_key,
            deps=None,
            fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
        ),
    ):
        pass

    class ParentV2(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["code_version", "parent_v2"]),
            deps=None,
            fields=[FieldSpec(key=FieldKey(["default"]), code_version=2)],
        ),
    ):
        pass

    class ChildWithParentV1(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["code_version", "child_v1"]),
            deps=[FeatureDep(key=parent_key)],
            fields=[FieldSpec(key=FieldKey(["default"]), code_version=7)],
        ),
    ):
        pass

    class ChildWithParentV2(
        Feature,
        spec=FeatureSpec(
            key=FeatureKey(["code_version", "child_v2"]),
            deps=[FeatureDep(key=ParentV2.spec.key)],
            fields=[FieldSpec(key=FieldKey(["default"]), code_version=7)],
        ),
    ):
        pass

    assert ChildWithParentV1.code_version == ChildWithParentV2.code_version


def test_code_version_deterministic_per_class(graph) -> None:
    feature_cls = _build_feature(
        key_suffix="deterministic",
        fields=[FieldSpec(key=FieldKey(["default"]), code_version=3)],
    )

    first = feature_cls.code_version
    second = feature_cls().code_version
    third = feature_cls().code_version

    assert first == second == third


@given(
    st.lists(
        st.integers(min_value=0, max_value=1_000_000),
        min_size=1,
        max_size=6,
    )
)
def test_code_version_consistent_across_field_order(
    graph, code_versions: list[int]
) -> None:
    field_items = [
        (FieldKey([f"field_{idx}"]), code_version)
        for idx, code_version in enumerate(code_versions)
    ]

    def build_fields(items: list[tuple[FieldKey, int]]) -> list[FieldSpec]:
        return [
            FieldSpec(key=field_key, code_version=code_version)
            for field_key, code_version in items
        ]

    unique_suffix = uuid4().hex
    ordered_feature = _build_feature(
        key_suffix=f"ordered_{unique_suffix}",
        fields=build_fields(field_items),
    )
    shuffled_feature = _build_feature(
        key_suffix=f"shuffled_{unique_suffix}",
        fields=build_fields(list(reversed(field_items))),
    )

    assert ordered_feature.code_version == shuffled_feature.code_version
