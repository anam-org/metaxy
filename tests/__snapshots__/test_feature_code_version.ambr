# serializer version: 1
# name: test_code_version_multiple_fields
  'f73ebda5b15e6a435a336e19fb7a9a6e7e2af79da7374b0fab08e4d0f2527f31'
# ---
# name: test_code_version_single_field
  '08d017a7f6a485735df6022f51f5b4a9c0b6fc3f0d4be1a2ead8bf9135a99993'
# ---
